File: scripts/truncate_descriptions.py
#!/usr/bin/env python3
"""Truncate JSON description fields to 155 characters with word boundary preservation."""

import json
import os
from pathlib import Path


def truncate_description(text: str, max_length: int = 155) -> str:
    """Truncate text to max_length, preserving word boundary and adding ellipsis."""
    if len(text) <= max_length:
        return text

    # Truncate to max_length - 3 to account for ellipsis
    truncated = text[:max_length - 3]

    # Find last space to preserve word boundary
    last_space = truncated.rfind(' ')

    if last_space > 0:
        truncated = truncated[:last_space]

    return truncated + '...'


def process_json_file(file_path: Path) -> tuple[bool, int]:
    """Process a single JSON file. Returns (was_modified, original_length)."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)

        if 'description' not in data:
            return False, 0

        original_desc = data['description']
        original_length = len(original_desc)

        if original_length <= 155:
            return False, original_length

        # Truncate and save
        data['description'] = truncate_description(original_desc)

        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)

        return True, original_length

    except Exception as e:
        print(f"Error processing {file_path}: {e}")
        return False, 0


def process_directory(directory: Path) -> dict:
    """Process all JSON files in a directory."""
    stats = {
        'total_files': 0,
        'modified_files': 0,
        'skipped_files': 0,
        'total_chars_saved': 0
    }

    if not directory.exists():
        print(f"Directory not found: {directory}")
        return stats

    json_files = list(directory.glob('*.json'))
    stats['total_files'] = len(json_files)

    for json_file in json_files:
        was_modified, original_length = process_json_file(json_file)

        if was_modified:
            stats['modified_files'] += 1
            stats['total_chars_saved'] += original_length - 155
        else:
            stats['skipped_files'] += 1

    return stats


def main():
    """Process all target directories."""
    base_dir = Path('/Users/gejiayu/owner/seo/data')

    directories = [
        'travel-hospitality-tools',
        'insurance-claims-processing-tools'
    ]

    total_stats = {
        'total_files': 0,
        'modified_files': 0,
        'skipped_files': 0,
        'total_chars_saved': 0
    }

    for dir_name in directories:
        print(f"\n{'='*60}")
        print(f"Processing: {dir_name}")
        print(f"{'='*60}")

        # Process main directory
        main_dir = base_dir / dir_name
        stats = process_directory(main_dir)

        print(f"Main directory: {stats['total_files']} files")
        print(f"  - Modified: {stats['modified_files']}")
        print(f"  - Skipped (already <= 155 chars): {stats['skipped_files']}")
        print(f"  - Characters saved: {stats['total_chars_saved']}")

        # Update totals
        for key in total_stats:
            total_stats[key] += stats[key]

        # Process zh/ counterpart
        zh_dir = base_dir / 'zh' / dir_name
        if zh_dir.exists():
            zh_stats = process_directory(zh_dir)
            print(f"\nZH directory: {zh_stats['total_files']} files")
            print(f"  - Modified: {zh_stats['modified_files']}")
            print(f"  - Skipped: {zh_stats['skipped_files']}")

            # Update totals
            for key in total_stats:
                total_stats[key] += zh_stats[key]

    print(f"\n{'='*60}")
    print("SUMMARY")
    print(f"{'='*60}")
    print(f"Total files processed: {total_stats['total_files']}")
    print(f"Total files modified: {total_stats['modified_files']}")
    print(f"Total files skipped: {total_stats['skipped_files']}")
    print(f"Total characters saved: {total_stats['total_chars_saved']}")


if __name__ == '__main__':
    main()

File: scripts/truncate-descriptions.js
const fs = require('fs');
const path = require('path');

const MAX_DESC_LENGTH = 155;

function truncateDescription(description) {
  if (!description) return description;
  if (description.length <= MAX_DESC_LENGTH) return description;

  // 找到155字符附近最近的空格（往前找），保留单词边界
  const truncated = description.substring(0, MAX_DESC_LENGTH);
  const lastSpace = truncated.lastIndexOf(' ');

  if (lastSpace > 0) {
    return truncated.substring(0, lastSpace) + '...';
  } else {
    return truncated + '...';
  }
}

function processFile(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const json = JSON.parse(content);

    const originalDesc = json.description;
    const truncatedDesc = truncateDescription(originalDesc);

    if (originalDesc !== truncatedDesc) {
      json.description = truncatedDesc;
      fs.writeFileSync(filePath, JSON.stringify(json, null, 2), 'utf8');
      return { processed: true, originalLength: originalDesc.length, newLength: truncatedDesc.length };
    }

    return { processed: false };
  } catch (error) {
    console.error(`Error processing ${filePath}:`, error.message);
    return { error: true };
  }
}

function processDirectory(directory) {
  const files = fs.readdirSync(directory).filter(file => file.endsWith('.json'));

  let stats = {
    total: files.length,
    processed: 0,
    skipped: 0,
    errors: 0
  };

  for (const file of files) {
    const filePath = path.join(directory, file);
    const result = processFile(filePath);

    if (result.error) {
      stats.errors++;
    } else if (result.processed) {
      stats.processed++;
    } else {
      stats.skipped++;
    }
  }

  return stats;
}

// 处理三个目录
const directories = [
  'data/customer-support-tools',
  'data/warehouse-inventory-tools',
  'data/transportation-fleet-tools'
];

const zhDirectories = [
  'data/zh/customer-support-tools',
  'data/zh/warehouse-inventory-tools',
  'data/zh/transportation-fleet-tools'
];

console.log('Processing English directories...\n');

let totalStats = {
  total: 0,
  processed: 0,
  skipped: 0,
  errors: 0
};

for (const dir of directories) {
  const stats = processDirectory(dir);
  console.log(`${dir}:`);
  console.log(`  Total: ${stats.total}`);
  console.log(`  Processed: ${stats.processed}`);
  console.log(`  Skipped: ${stats.skipped}`);
  console.log(`  Errors: ${stats.errors}\n`);

  totalStats.total += stats.total;
  totalStats.processed += stats.processed;
  totalStats.skipped += stats.skipped;
  totalStats.errors += stats.errors;
}

console.log('Processing Chinese (zh) directories...\n');

for (const dir of zhDirectories) {
  if (fs.existsSync(dir)) {
    const stats = processDirectory(dir);
    console.log(`${dir}:`);
    console.log(`  Total: ${stats.total}`);
    console.log(`  Processed: ${stats.processed}`);
    console.log(`  Skipped: ${stats.skipped}`);
    console.log(`  Errors: ${stats.errors}\n`);

    totalStats.total += stats.total;
    totalStats.processed += stats.processed;
    totalStats.skipped += stats.skipped;
    totalStats.errors += stats.errors;
  } else {
    console.log(`${dir}: Directory not found\n`);
  }
}

console.log('====================');
console.log('Total Summary:');
console.log(`  Total files: ${totalStats.total}`);
console.log(`  Processed: ${totalStats.processed}`);
console.log(`  Skipped: ${totalStats.skipped}`);
console.log(`  Errors: ${totalStats.errors}`);